### DISCUSSION TITLE
Ice test demo

### DISCUSSION THREAD

**Discussion by kexiongwang** [zeroc-ice/ice#3260] [Q&A]
Q1:whats difference between test in the cpp folder of ice and test in ice-demo?
Q2: i have make install ice-3.7.1,but test in the cpp folder of ice ,allTest.py runs failure , i check the slice,however ,i cant find the error 
![image](https://github.com/user-attachments/assets/b6a24cfe-2c8f-47aa-a068-0ca43a50f6e9)
Comment: update gcc 4.8.5 to gcc 9.4

**Comment by pepone** [comment]
> Q1:whats difference between test in the cpp folder of ice and test in ice-demo?

The tests under `ice/cpp/test` are for the C++ language mapping itself. These are designed to verify the implementation of Ice for C++. On the other hand, ice-demos does not contain tests. Instead, it provides sample applications that demonstrate how to use the C++ language mapping. Specifically, you’ll find these examples for the C++11 mapping under `ice-demos/cpp11` or the samples for the C++98 mapping under `ice-demos/cpp98`. Note that these samples are located in the 3.7 branch of the ice-demos repository, which corresponds to the Ice 3.7 series.

> i have make install ice-3.7.1,but test in the cpp folder of ice ,allTest.py runs failure , i check the slice,however ,i cant find the error

- What OS are you using?
- Have you build MCPP yourself or are you using a package?

**Reply by kexiongwang** [reply]
OS:redhat 7.8
build MCPP
![image](https://github.com/user-attachments/assets/fbdf33de-8e18-4225-953c-0d42d56f4cc8)

**Reply by pepone** [reply]
The failing test is verifying that slice2cpp reports the expected errors. There seems to be a mismatch here, so examining the actual reported errors can help us figure this out.

Could you run the following command in your 3.7.10 build and paste the output?

```
./bin/slice2cpp -I. ./test/Slice/errorDetection/CaseInsensitive.ice
```

**Reply by kexiongwang** [reply]
![微信图片_20241217083928](https://github.com/user-attachments/assets/9854f255-61ed-4f4b-a7fa-72bc48fd1fa0)

i make install ice-3.7.1

**Comment by pepone** [comment]
The output looks correct. It is not clear why it is failing when you run the test.

Can you try to upgrade the `test/Slice/errorDetection/test.py` script as in:

```
--- a/cpp/test/Slice/errorDetection/test.py
+++ b/cpp/test/Slice/errorDetection/test.py
@@ -36,7 +36,7 @@ class SliceErrorDetectionTestCase(ClientTestCase):
                 with open(os.path.join(testdir, regex1.sub(".err", file)), "r") as f:
                     lines2 = f.readlines()
                     if len(lines1) != len(lines2):
-                        raise RuntimeError("failed (lines1 = {0}, lines2 = {1})!".format(len(lines1), len(lines2)))
+                        raise RuntimeError("failed (lines1 = {0}, lines2 = {1})!\n{2}".format(len(lines1), len(lines2), lines1))
 
```

And then run again the test with:

```
python3 allTests.py Slice/errorDetection --debug
```

**Reply by kexiongwang** [reply]
allTests.py path /opt/ice/cpp
Slice/errorDetection path /opt/ice/cpp/testa
i make install Ice3.7.1 in /opt/Ice3.7.1
![d1632b588d58559d98069487716bc9e](https://github.com/user-attachments/assets/1c9fda75-48c5-4346-bf71-9d500299187e)
![Uploading 微信图片_20241217170850.jpg…]()

another question, we use python2.7 for the first choice, higher version is testing

**Reply by pepone** [reply]
Seems your editor inserted a `tab` character when updating the script, python doesn't like this. You have to replace it to use space for indentation.

Using a different Python version for the C++ mapping tests should be fine. 

For the Python mapping tests you need to use the interpreter for which the extension was built. if that is not the same as the one you are using to run the tests script you can change it with `--python` option

```
Python mapping options:
--python=<interpreter>   Choose the interpreter used to run python tests
```

**Reply by kexiongwang** [reply]
![ice3 7 1_allTest py](https://github.com/user-attachments/assets/e4a69777-cf8d-4396-b522-a6c01445cb36)

 i change python version , it seems work ,however , works a little
i make a gcc-9.4 version, and i change the enviroment ,i put gcc-9.4 bin befor system gcc4.8.5

**Reply by pepone** [reply]
Here the compiler is failing to run because it cannot load the C++ runtime "libstdc++".

This typically happens when the compiler was built with an incompatible toolchain. Are you running the tests in the same machine that you build Ice or is this a different machine?

**Reply by kexiongwang** [reply]
same machine ,and i try change the enviroment

**Reply by pepone** [reply]
Seems there is a problem with the installation of your libstd++ for GCC 9.4.

**Reply by kexiongwang** [reply]
you're right, it involves system kernel ,i should be careful ,i change the libstdc++.so link, it works ,however ,it still remains a small problem 
![eb1c64207b63824697d01269d49bd4b](https://github.com/user-attachments/assets/cb5150d8-4a12-4713-a4bf-b3a1e47adf9f)
![d931c85519d2e79adfe5406a7171a43](https://github.com/user-attachments/assets/096bc0e5-f3a2-44d3-86af-cf52c7f5ac90)

**Reply by pepone** [reply]
> i change the libstdc++.so link

That sounds like a bad Idea there should be a way to properly install G++ 9 toolchain in rhel7

Also is there any reason for using 3.7.1? It is better to use last patch release 3.7.10 and one of the supported platforms.

https://doc.zeroc.com/ice/3.7/release-notes/supported-platforms-for-ice-3-7-10

**Reply by kexiongwang** [reply]
we use Ice3.7.1 for long time , it's hard to make change in a short time , so i have to slove the question. 
you means the error is due to Ice3.7.1 runs in redhat 7.8 and  gcc9.4 ?

**Reply by pepone** [reply]
> we use Ice3.7.1 for long time , it's hard to make change in a short time 

I 3.7.10 is binary compatible with 3.7.1, it should't require updates to your application.

> you means the error is due to Ice3.7.1 runs in redhat 7.8 and gcc9.4 ?

No. The error seems related to your gcc9.4 setup. We don't support this configuration (Red Hat 7 + GCC 9.4) it might work but it is something we don't test or support.

**Reply by kexiongwang** [reply]
ok ,thanks!
